### ISSUE TITLE
Can't use debian DOD with / in component

### ISSUE THREAD

**Issue by dbnicholson** [openSUSE/open-build-service#5217]
I'm trying to setup a DOD project for debian security. Unfortunately, the security repo uses `/` in component names as seen in http://security.debian.org/dists/stretch/updates/Release. This messes with the URL heuristics in `bs_dodup` since it thinks everything after the last `/` is the component.

Expected Result
============
The doddata is updated.

How to Reproduce
===============
I don't have a public server with this, and you wouldn't be able to see anything in the web UI. Here's the project meta I'm using:
```
<project name="debian:stretch:security">
  <title>Debian stretch security DOD</title>
  <description>Download on demand from Debian stretch security</description>
  <person userid="admin" role="bugowner"/>
  <person userid="admin" role="maintainer"/>
  <build>
    <disable/>
  </build>
  <publish>
    <disable/>
  </publish>
  <repository name="main">
    <download arch="i586" url="http://security.debian.org/stretch/updates/main" repotype="deb"/>
    <download arch="x86_64" url="http://security.debian.org/stretch/updates/main" repotype="deb"/>
    <download arch="armv7hl" url="http://security.debian.org/stretch/updates/main" repotype="deb"/>
    <path project="debian:stretch:updates" repository="main"/>
    <path project="debian:stretch" repository="main"/>
    <arch>i586</arch>
    <arch>x86_64</arch>
    <arch>armv7hl</arch>
  </repository>
</project>
```

Further Information
===============
* I'm running 2.7.4, but I don't see any recent changes in `bs_dodup` that would change this.

**Comment by dbnicholson** [comment]
Would you consider adding some real attributes to the `<download>` node like `dist="stretch" components="updates/main"`? Otherwise, any suggestions for mangling the exist parsing?

**Comment by dbnicholson** [comment]
I said that backwards, actually. The dist (`stretch/updates`) has a slash in it. Maybe there's a simpler bug fix for that.

**Comment by mlschroe** [comment]
Yes, I'd love to use query parameters. Is there some official way to represent a debian repo as url?

**Comment by dbnicholson** [comment]
The hack I came up with now is to separate the distribution and components by `|` in the "url". E.g., I changed my URL to `http://security.debian.org|stretch/updates|main`. Here's the patch to handle that:
```
--- bs_dodup.orig	2018-06-21 07:28:33.000000000 -0700
+++ bs_dodup	2018-06-26 05:00:53.200552825 -0700
@@ -218,6 +221,15 @@
     @components = ('.');
     $url = defined($2) ? "$1$2" : $1;
     $url .= '/' unless $url =~ /\/$/;
+  } elsif ($url =~ /\|/) {
+    # Dist and components split by |
+    my @parts = split(/\|/, $url);
+    $baseurl = shift @parts;
+    my $dist = shift @parts;
+    @components = @parts;
+    push @components, 'main' unless @components;
+    $baseurl .= '/' unless $baseurl =~ /\/$/;
+    $url = "${baseurl}dists/${dist}/";
   } else {
     if ($url =~ /([^\/]+)$/) {
       @components = split(/[,+]/, $1);
```
What do you think about that? I can look at using query parameters, though.

**Comment by dbnicholson** [comment]
I made a PR in #5223 for using `|` delimited URLs. I also played around with using a query string and parsing the URL using `URI::URL`. It worked out fine, although it was a little awkward when putting the URL in the project metadata because the `&` in the query string needs to be converted to `&amp;` to be legal XML. I can make a PR for that if you prefer.

**Comment by dbnicholson** [comment]
PR #5225 parses the distro and components from a query string.

**Comment by bluca** [comment]
@dbnicholson there's another issue with DoDs that I just ran into. Consider the libc6 case for Stretch for example:

```
$ apt-cache policy libc6
libc6:
  Installed: 2.24-11+deb9u3
  Candidate: 2.24-11+deb9u3
  Version table:
 *** 2.24-11+deb9u3 500
        500 https://deb.debian.org/debian stretch/main amd64 Packages
        100 /var/lib/dpkg/status
     2.24-11+deb9u1 500
        500 http://security.debian.org/debian-security stretch/updates/main amd64 Packages
```

At the moment, the latest version comes from the "stretch" repository, as a non-security critical issue was fixed through proposed-updates and then added to the "stretch" repository with the last 9.x point release.
The "security" repository holds the last version that had a CVE (or otherwise security-related) fix, which is older than what was added to the latest point release.
With apt this works fine, as it will select the highest version.

But in OBS a DoD can only reference one repository - so you get 3 targets, like you have in your meta example - one for stable, one for updates and one for security.
OBS will, by design I think, not pick the highest version, but the first instance of the required package it finds while walking the list of targets top-down.
So if the security DoD comes first in the meta for the target, then an older version of libc gets picked. But of course if security comes last, then all security updates get ignored, since all packages that are updated in security were part of the original release.

As far as I can see, the only way would be to change the DoD implementation to allow multiple "download" elements of the same architecture, and then de-duplicating them by picking the highest version locally before publishing.

Thoughts?

**Comment by dbnicholson** [comment]
I believe this can be closed since #5225 (or rather #6574) was merged.

**Comment by bluca** [comment]
> I believe this can be closed since #5225 (or rather #6574) was merged.

Great, thanks for working on that!

Have you had a chance to look at my previous comment? I think it's a major issue, and I don't think dod can be used with Debian/Ubuntu repositories in production if we don't find a solution, unfortunately

**Comment by dbnicholson** [comment]
@bluca Sorry for not responding earlier. That is a legitimate concern and one I hit in practice a couple times when something in security (which is first in our path) masks a newer version in updates (which is second in our path). I'm not sure there's a lot you can do about it with the existing setup because (I believe) OBS will always take the first version found in referenced repositories.

On Ubuntu all of the dists come from the same URL, so you could possibly rework DODs to merge multiple dists into one repo, preferring whatever version is newest. However, on Debian security comes from a different URL, so you'd need an entirely different configuration scheme to do this merging scheme.

That said, I don't find it to be that big of a deal really. Once something newer goes into updates, then the version in security is dropped. So, the issue is resolved by way of the upstream policy after a bit. It's pretty rare that you'd require a build dep that just landed in updates and is currently masked by something in security.

That's just how Debian and Ubuntu do it, though. Generally you can get into situations with OBS where a newer dependency further down the path is masked. I think the general fix is to have a repo option that says to resolve the newest versions from all repos in the path rather than the first found.

**Comment by bluca** [comment]
> [@dbnicholson](https://github.com/dbnicholson) there's another issue with DoDs that I just ran into. Consider the libc6 case for Stretch for example:
> 
> ```
> $ apt-cache policy libc6
> libc6:
>   Installed: 2.24-11+deb9u3
>   Candidate: 2.24-11+deb9u3
>   Version table:
>  *** 2.24-11+deb9u3 500
>         500 https://deb.debian.org/debian stretch/main amd64 Packages
>         100 /var/lib/dpkg/status
>      2.24-11+deb9u1 500
>         500 http://security.debian.org/debian-security stretch/updates/main amd64 Packages
> ```
> 
> At the moment, the latest version comes from the "stretch" repository, as a non-security critical issue was fixed through proposed-updates and then added to the "stretch" repository with the last 9.x point release. The "security" repository holds the last version that had a CVE (or otherwise security-related) fix, which is older than what was added to the latest point release. With apt this works fine, as it will select the highest version.
> 
> But in OBS a DoD can only reference one repository - so you get 3 targets, like you have in your meta example - one for stable, one for updates and one for security. OBS will, by design I think, not pick the highest version, but the first instance of the required package it finds while walking the list of targets top-down. So if the security DoD comes first in the meta for the target, then an older version of libc gets picked. But of course if security comes last, then all security updates get ignored, since all packages that are updated in security were part of the original release.
> 
> As far as I can see, the only way would be to change the DoD implementation to allow multiple "download" elements of the same architecture, and then de-duplicating them by picking the highest version locally before publishing.
> 
> Thoughts?

I have a fix for this now: https://github.com/openSUSE/open-build-service/pull/19085
